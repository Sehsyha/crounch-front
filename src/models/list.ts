import { Product } from './product';
import { CategoryInGetListResponse } from './category';

interface List {
  id: string
  name: string
  products: Product[]
}

interface GetListResponse {
  id: string
  name: string
  creationDate: string
  categories: CategoryInGetListResponse[]
}

export { List, GetListResponse }
