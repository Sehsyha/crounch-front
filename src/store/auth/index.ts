import { Module } from 'vuex';

import { TOKEN_STORAGE_KEY } from '@/utils/constants';
import { getters } from './getters';
import { actions } from './actions';
import { mutations } from './mutations';
import { RootState } from '..';

export interface AuthState {
  token: string;
  status: string;
}

export const state: AuthState = {
  token: localStorage.getItem(TOKEN_STORAGE_KEY) || '',
  status: '',
};

const namespaced: boolean = true;
export const authNamespace = { namespace: 'auth' };

export const auth: Module<AuthState, RootState> = {
  namespaced,
  state,
  getters,
  actions,
  mutations,
};
