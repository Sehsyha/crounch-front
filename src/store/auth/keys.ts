export enum AuthActions {
  LOGIN = 'LOGIN',
  SIGNUP = 'SIGNUP',
  LOGOUT = 'LOGOUT',
}

export enum AuthGetters {
  IS_AUTHENTICATED = 'IS_AUTHENTICATED',

}

export enum AuthMutations {
  LOGIN = 'LOGIN',
  LOGOUT = 'LOGOUT',
}
