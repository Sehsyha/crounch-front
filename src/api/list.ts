import { doFetch } from './api';
import { List } from '@/models/list';

export function createList(name: string): Promise<List> {
  return doFetch({
    url: 'lists',
    method: 'POST',
    data: { name }
  });
}

export function getOwnerLists(): Promise<Array<List>> {
  return doFetch({
    url: 'lists',
    method: 'GET',
  });
}

export function deleteList(id: string): Promise<void> {
  return doFetch({
    url: `lists/${id}`,
    method: 'DELETE',
  });
}

export function addProductToList(productID: string, listID: string): Promise<void> {
  return doFetch({
    url: `lists/${listID}/products/${productID}`,
    method: 'POST'
  })
}

export function readList(listID: string): Promise<List> {
  return doFetch({
    url: `lists/${listID}`,
    method: 'GET'
  })
}
