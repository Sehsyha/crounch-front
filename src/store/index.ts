import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import { auth } from './auth';
import { ListModule } from './ListModule';
import { product } from './product'

Vue.use(Vuex);

export interface RootState { }

const store: StoreOptions<RootState> = {
  state: {},
  modules: {
    auth,
    ListModule,
    product,
  },
  strict: true,
};

export default new Vuex.Store<RootState>(store);
