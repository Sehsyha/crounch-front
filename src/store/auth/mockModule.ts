import { AuthKeys } from './keys';
import { Module } from 'vuex';
import { AuthState } from '.';
import { RootState } from '..';

const actions = {
  [AuthKeys.SIGNUP]: jest.fn(),
  [AuthKeys.LOGIN]: jest.fn(),
  [AuthKeys.LOGOUT]: jest.fn(),
};

const getters = {
  isAuthenticated: jest.fn(),
};

export function createAuthModuleMock(): Module<AuthState, RootState> {
  return {
    namespaced: true,
    actions,
    getters,
    state: {
      token: '',
      status: '',
    },
  };
}
