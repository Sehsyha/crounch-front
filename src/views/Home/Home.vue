<template>
  <div class="home">
    <HelloWorld msg="Do you want to Crounch ?" />
    <h1 v-if="this.isAuthenticated">
      Connecté
      <a id="logout" @click.prevent="logout">Se déconnecter</a>
    </h1>
    <Signup v-show="!this.isAuthenticated" />
    <Login v-show="!this.isAuthenticated" />
    <Search />
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import { Getter, Action } from 'vuex-class';

import HelloWorld from '@/components/HelloWorld/HelloWorld.vue';
import Signup from '@/components/Signup/Signup.vue';
import Login from '@/components/Login/Login.vue';
import Search from '@/components/Search/Search.vue';
import { authNamespace } from '../../store/auth';
import { AuthKeys } from '../../store/auth/keys';

@Component({
  components: {
    HelloWorld,
    Signup,
    Login,
    Search,
  },
})
export default class Home extends Vue {
  @Getter('isAuthenticated', authNamespace) isAuthenticated!: boolean;
  @Action(AuthKeys.LOGOUT, authNamespace) doLogout: any;

  logout() {
    this.doLogout();
  }
}
</script>
