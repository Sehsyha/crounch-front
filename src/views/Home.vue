<template>
  <div class="home">
    <HelloWorld msg="Voulez-vous crouncher avec moi ce soir ?" />
    <Signup />
    <Login />
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import { getModule } from 'vuex-module-decorators';

import HelloWorld from '@/components/HelloWorld.vue';
import Signup from '@/components/Signup.vue';
import Login from '@/components/Login.vue';
import CreateList from '@/components/CreateList.vue';
import { ListModule } from '@/store/list/ListModule';
import { AuthModule } from '@/store/auth/AuthModule';

@Component({
  components: {
    HelloWorld,
    Signup,
    Login,
  },
})
export default class Home extends Vue {
  public authModule: AuthModule = getModule(AuthModule)

  mounted() {
    if (this.authModule.isAuthenticated) {
      this.$router.push('/lists');
    }
  }
}
</script>
