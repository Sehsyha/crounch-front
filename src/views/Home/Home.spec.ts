import { Wrapper } from '@vue/test-utils';
import { Module } from 'vuex';

import { shallowComponent } from '../../utils/test';
import Home from './Home.vue';
import { AuthState } from '@/store/auth';
import { RootState } from '@/store';
import { createAuthModuleMock } from '@/store/auth/mockModule';
import { AuthKeys } from '@/store/auth/keys';
import { ListKeys } from '@/store/list/keys';
import { createListModuleMock } from '@/store/list/mockModule';
import { ListState } from '@/store/list';

describe('Home', () => {
  let wrapper: Wrapper<Home>;
  let auth: Module<AuthState, RootState>;
  let list: Module<ListState, RootState>;

  beforeEach(() => {
    auth = createAuthModuleMock();
    list = createListModuleMock();

    const modules = {
      auth,
      list,
    };

    wrapper = shallowComponent(Home, { modules });
  });

  it('Should render.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render when user is authenticated.', () => {
    auth = createAuthModuleMock();
    (auth.getters!['isAuthenticated'] as jest.Mock).mockReturnValue(true);

    const modules = {
      auth,
    };

    wrapper = shallowComponent(Home, { modules });

    expect(wrapper).toMatchSnapshot();
  });

  it('Should call logout when logout button is clicked.', () => {
    wrapper.find('#logout').trigger('click');

    expect(auth.actions![AuthKeys.LOGOUT]).toHaveBeenCalled();
  });

  it('Should redirect to lists page on creation when user is authenticated.', () => {
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith('lists');
  });

  it('Should reset the lists in state.', () => {
    wrapper.find('#logout').trigger('click');

    expect(list.mutations![ListKeys.RESET]).toHaveBeenCalled();
  });
});
