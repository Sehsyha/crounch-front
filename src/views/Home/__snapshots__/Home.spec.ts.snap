// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Should render when user is authenticated. 1`] = `
<div class="home">
  <helloworld-stub msg="Voulez-vous crouncher avec moi ce soir ?"></helloworld-stub>
  <h1>Connect√©</h1>
  <signup-stub style="display: none;"></signup-stub>
  <login-stub style="display: none;"></login-stub>
</div>
`;

exports[`Home Should render when user is not authenticated. 1`] = `
<div class="home">
  <helloworld-stub msg="Voulez-vous crouncher avec moi ce soir ?"></helloworld-stub>
  <!---->
  <signup-stub></signup-stub>
  <login-stub></login-stub>
</div>
`;
